File: download.py
import requests
import zipfile
import os

url = "https://data.nationalgrideso.com/system/system-frequency-data/datapackage.json"
download_to = "tmp/"
extract_to = "data/"

resources = requests.get(url).json()["resources"]

print(f"Found {len(resources)} datasets")

for ix, resource in enumerate(resources):
    path = os.path.join(download_to, f"{resource['id']}.zip");
    print(f"[{ix:02}/{len(resources)}] Saving '{resource['title']:<40}' to {path}")

    response = requests.get(resource["path"])
    with open(path, "wb") as f:
        f.write(response.content)

for ix, filename in enumerate(os.listdir(download_to)):
    path = os.path.join(download_to, filename)
    print(f"[{ix:02}/{len(resources)}] Inflating '{path}'")

    try:
        with zipfile.ZipFile(path, "r") as zip_ref:
            zip_ref.extractall(extract_to)
    except:
        print(f"Failed to unpack {path}")

File: src/download.rs
use crate::cli::Cli;
use anyhow::Result;
use log::*;
use reqwest;
use serde::Deserialize;
use std::io::Read;

#[derive(Deserialize)]
struct Response {
    resources: Vec<Resource>,
}

#[derive(Deserialize)]
struct Resource {
    title: String,
    path: String,
}

/// Download the dataset
pub fn download(args: &Cli) -> Result<()> {
    let client = reqwest::blocking::Client::new();

    // Initial request, get the months where data is available
    let response: Response = client
        .get("https://data.nationalgrideso.com/system/system-frequency-data/datapackage.json")
        .send()?
        .json()?;

    // Save every resource
    log::info!("Starting download");
    for (ix, resource) in response.resources.iter().enumerate() {
        log::info!(
            "[{:02}/{}] Saving '{:<40}' to {:?}",
            ix,
            response.resources.len(),
            resource.title,
            args.data_path
        );
        let bytes = client.get(&resource.path).send()?.bytes()?;

        // wrap the byte array in a Cursor, providing it with a Seek implementation
        let mut zip = zip::ZipArchive::new(std::io::Cursor::new(bytes))?;
        zip.extract(&args.data_path)?;
    }

    log::info!("Finished downloading the dataset");

    Ok(())
}
